zerver/worker/queue_processors.py

```diff
@@ -115,25 +115,77 @@ def on_failure(event: Dict[str, Any]) -> None:
         if self.queue_name is None:
             raise WorkerDeclarationException("Queue worker declared without queue_name")
+        self.queue_last_emptied_timestamp = time.time()
+        self.consumed_since_last_emptied = 0
+        self.recent_consume_times = deque(maxlen=50)  # type: MutableSequence[Tuple[int, float]]
+        self.consume_interation_counter = 0
 
     @abstractmethod
     def consume(self, data: Dict[str, Any]) -> None:
         pass
 
-    def consume_wrapper(self, data: Dict[str, Any]) -> None:
+    def do_consume(self, consume_func: Callable[[List[Dict[str, Any]]], None],
+                   events: List[Dict[str, Any]]) -> None:
         try:
-            self.consume(data)
+            time_start = time.time()
+            consume_func(events)
+            consume_time_seconds = time.time() - time_start  # type: Optional[float]
+            self.consumed_since_last_emptied += len(events)
         except Exception:
-            self._handle_consume_exception([data])
+            self._handle_consume_exception(events)
+            consume_time_seconds = None
         finally:
             flush_per_request_caches()
             reset_queries()
 
+            if consume_time_seconds is not None:
+                self.recent_consume_times.append((len(events), consume_time_seconds))
+
+            self.consume_interation_counter += 1
+            if self.consume_interation_counter >= self.CONSUME_ITERATIONS_BEFORE_UPDATE_STATS_NUM:
+                remaining_queue_size = self.q.queue_size(self.queue_name)
+                if remaining_queue_size == 0:
+                    self.queue_last_emptied_timestamp = time.time()
+                    self.consumed_since_last_emptied = 0
+
+                self.consume_interation_counter = 0
+                self.update_statistics(remaining_queue_size)
+
+    def update_statistics(self, remaining_queue_size: int) -> None:
+        total_seconds = sum([seconds for _, seconds in self.recent_consume_times])
+        total_events = sum([events_number for events_number, _ in self.recent_consume_times])
+        if total_events == 0:
+            recent_average_consume_time = None
+        else:
+            recent_average_consume_time = total_seconds / total_events
+        stats_dict = dict(
+            update_time=time.time(),
+            recent_average_consume_time=recent_average_consume_time,
+            current_queue_size=remaining_queue_size,
+            queue_last_emptied_timestamp=self.queue_last_emptied_timestamp,
+            consumed_since_last_emptied=self.consumed_since_last_emptied,
+        )
+
+        if not os.path.exists(settings.QUEUE_STATS_DIR):
+            os.mkdir(settings.QUEUE_STATS_DIR)  # nocoverage
+
+        fname = '%s.stats' % (self.queue_name,)
+        fn = os.path.join(settings.QUEUE_STATS_DIR, fname)
+        lock_fn = fn + '.lock'
+        with lockfile(lock_fn):
+            with open(fn, 'w') as f:
```

Comment:
Is the idea that client reading this will also grab the lock?  I guess that is an alternative to writing to a `.tmp` file and then moving it over using `mv` (which also ensures atomic update).  Personally I think the atomic update approach might be better.

Reply:
I was thinking of something like that yes, plus `_handle_consume_exception` uses this pattern, what's the reason for it there?

Reply:
Ah, I think I know why, `test-backend` may have these running in parallel.